### PR TITLE
Implementation IPresenterFormatter

### PR DESCRIPTION
**Pull request by Trejjam**
Based on https://github.com/nette/di/pull/104#issuecomment-204951354

### CODE REVIEW

File: src/Application/PresenterFormatter/PresenterFormatterManager.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ */
+
+namespace Nette\Application\PresenterFormatter;
+
+use Nette\Application\PresenterFormatterException;
+
+/**
+ * Responsible for manage multiple IPresenterFormatter
+ */
+class PresenterFormatterManager implements IPresenterFormatter
```
Comment (Majkl578) [inline src/Application/PresenterFormatter/PresenterFormatterManager.php]: This is typically named Chain, so e.g. PresenterFormatterChain, with addFormatter method.

**Comment by dg** [comment]
https://forum.nette.org/cs/26116-presentermapper-ipresenterformatter-apod

File: src/Application/PresenterFactory.php [hunk 2]
```diff
@@ -17,31 +17,31 @@ class PresenterFactory implements IPresenterFactory
...
+	private $presenterFormatterManager;
+
 
 	/**
-	 * @param  callable  function (string $class): IPresenter
+	 * @param callable            $factory function (string $class): IPresenter
+	 * @param PresenterFormatter\PresenterFormatterManager $presenterFormatterManager
 	 */
-	public function __construct(callable $factory = NULL)
+	public function __construct(callable $factory = NULL, PresenterFormatter\PresenterFormatterManager $presenterFormatterManager)
```
Comment (fprochazka) [inline src/Application/PresenterFactory.php]: this should require the interface, not the chain implementation

File: src/Application/PresenterFactory.php [hunk 3]
```diff
@@ -17,31 +17,31 @@ class PresenterFactory implements IPresenterFactory
...
-	];
-
 	/** @var array */
 	private $cache = [];
 
 	/** @var callable */
 	private $factory;
 
+	/** @var PresenterFormatter\PresenterFormatterManager */
+	private $presenterFormatterManager;
```
Comment (fprochazka) [inline src/Application/PresenterFactory.php]: `presenterFormatter` is enough imho

**Comment by fprochazka** [comment]
This is gonna need a bit more work (mainly CS problems), but good job @Trejjam :+1:

File: src/Bridges/ApplicationDI/ApplicationExtension.php [hunk 4]
```diff
@@ -87,6 +95,7 @@ public function loadConfiguration()
 
 		if ($this->name === 'application') {
 			$container->addAlias('application', $this->prefix('application'));
+			$container->addAlias('nette.presenterFormatterChain', $this->prefix('presenterFormatterChain'));
```
Comment (fprochazka) [inline src/Bridges/ApplicationDI/ApplicationExtension.php]: imho this is for BC, so you do not have to create new aliases

File: src/Bridges/ApplicationDI/ApplicationExtension.php [hunk 5]
```diff
@@ -97,6 +106,18 @@ public function beforeCompile()
 		$container = $this->getContainerBuilder();
 		$all = [];
 
+		$presenterFormatterChain = $container->getDefinition($this->prefix('presenterFormatterChain'));
+		$defaultFormatter=$container->getDefinition($this->prefix('presenterFormatter.default'));
+
+		foreach ($container->findByType(Nette\Application\PresenterFormatter\IPresenterFormatter::class) as $presenterFormatter) {
```
Comment (fprochazka) [inline src/Bridges/ApplicationDI/ApplicationExtension.php]: I'm not sure this autodiscovery is a good thing

File: src/Bridges/ApplicationDI/ApplicationExtension.php (hunk 4 above)
Comment (Trejjam) [inline src/Bridges/ApplicationDI/ApplicationExtension.php]: Good to know, thanks

File: src/Bridges/ApplicationDI/ApplicationExtension.php (hunk 5 above)
Comment (Trejjam) [inline src/Bridges/ApplicationDI/ApplicationExtension.php]: Good point, it isn't good. I'll add it into config

**Pull request closed without merging**